smarty: double sets of quotes not working at end

The following example renders incorrectly:
``` py
import markdown

text = '''
He replied, "She said 'Hello.'"
'''

html = markdown.markdown(text, extensions=['smarty'])
print(html)
```

The output is:
```
<p>He replied, &ldquo;She said &lsquo;Hello.&rsquo;&ldquo;</p>
```

**The `&ldquo;` at the end of this example should be `&rdquo;`.**

The quote marks were replaced in this order:
1. openingSingleQuotesRegex
2. closingSingleQuotesRegex
3. openingDoubleQuotesRegex
4. remainingDoubleQuotesRegex

Unfortunately:
* closingDoubleQuotesRegex didn't match, because the `"` is not followed by a space (it's the last character of the string, which is stripped prior to replacement).
* closingDoubleQuotesRegex2 didn't match, because of the closing single quote replacement (i.e., the intermediate data in `SubstituteTextPattern.handleMatch()` contains `\u0003` before the `"`).

One possible way to fix this issue might be to change closingDoubleQuotesRegex to `r'"(?=\s|$)'` (i.e., double quote followed by whitespace *or end of string*). However that doesn't work in general, so a special case is likely needed.

> I don't think a word boundary check would be correct, but maybe a not word check `(?=\s|\W|$)`. It seems to solve the the HTML case, the non-word case, and the end of buffer case without breaking any tests.

Hm, what exactly change did you try? I have just tried this:
```diff
--- a/markdown/extensions/smarty.py
+++ b/markdown/extensions/smarty.py
@@ -142,7 +142,7 @@ decadeAbbrRe = r"(?<!\w)'(?=\d{2}s)"
 openingDoubleQuotesRegex = r'%s"(?=\w)' % openingQuotesBase
 
 # Double closing quotes:
-closingDoubleQuotesRegex = r'"(?=\s)'
+closingDoubleQuotesRegex = r'"(?=\s|\W|$)'
 closingDoubleQuotesRegex2 = '(?<=%s)"' % closeClass
 
 # Get most opening single quotes:
```

but it does break one of the tests:
```
======================================================================
FAIL: test_basic (tests.test_syntax.extensions.test_smarty.TestSmarty.test_basic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmitry/upstream/python-markdown/tests/test_syntax/extensions/test_smarty.py", line 59, in test_basic
    self.assertMarkdownRenders(
    ~~~~~~~~~~~~~~~~~~~~~~~~~~^
        '"[Link](http://example.com)" --- she said.',
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        '<p>&ldquo;<a href="http://example.com">Link</a>&rdquo; &mdash; she said.</p>'
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/home/dmitry/upstream/python-markdown/markdown/test_tools.py", line 77, in assertMarkdownRenders
    self.assertMultiLineEqual(output, expected)
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^
AssertionError: '<p>&rdquo;<a href="http://example.com">Link</a>&rdquo; &mdash; she said.</p>' != '<p>&ldquo;<a href="http://example.com">Link</a>&rdquo; &mdash; she said.</p>'
- <p>&rdquo;<a href="http://example.com">Link</a>&rdquo; &mdash; she said.</p>
?     ^
+ <p>&ldquo;<a href="http://example.com">Link</a>&rdquo; &mdash; she said.</p>
?     ^
```

--- a/docs/changelog.md
+++ b/docs/changelog.md
@@ -27,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Ensure `<center>` is treated like a block-level element (#1481).
 * Ensure that `abbr` extension respects `AtomicString` and does not process
   perceived abbreviations in these strings (#1512).
+* The `smarty` extension correctly renders nested closing quotes (#1514).
 
 ## [3.7] -- 2024-08-16
 

--- a/markdown/extensions/smarty.py
+++ b/markdown/extensions/smarty.py
@@ -134,6 +134,8 @@ doubleQuoteStartRe = r'^"(?=%s\B)' % punctClass
 #   <p>He said, "'Quoted' words in a larger quote."</p>
 doubleQuoteSetsRe = r""""'(?=\w)"""
 singleQuoteSetsRe = r"""'"(?=\w)"""
+doubleQuoteSetsRe2 = r'(?<=%s)\'"' % closeClass
+singleQuoteSetsRe2 = r"(?<=%s)\"'" % closeClass
 
 # Special case for decade abbreviations (the '80s):
 decadeAbbrRe = r"(?<!\w)'(?=\d{2}s)"
@@ -143,7 +145,7 @@ openingDoubleQuotesRegex = r'%s"(?=\w)' % openingQuotesBase
 
 # Double closing quotes:
 closingDoubleQuotesRegex = r'"(?=\s)'
-closingDoubleQuotesRegex2 = '(?<=%s)"' % closeClass
+closingDoubleQuotesRegex2 = r'(?<=%s)"' % closeClass
 
 # Get most opening single quotes:
 openingSingleQuotesRegex = r"%s'(?=\w)" % openingQuotesBase
@@ -240,6 +242,8 @@ class SmartyExtension(Extension):
             (doubleQuoteStartRe, (rdquo,)),
             (doubleQuoteSetsRe, (ldquo + lsquo,)),
             (singleQuoteSetsRe, (lsquo + ldquo,)),
+            (doubleQuoteSetsRe2, (rsquo + rdquo,)),
+            (singleQuoteSetsRe2, (rdquo + rsquo,)),
             (decadeAbbrRe, (rsquo,)),
             (openingSingleQuotesRegex, (1, lsquo)),
             (closingSingleQuotesRegex, (rsquo,)),

--- a/tests/test_syntax/extensions/test_smarty.py
+++ b/tests/test_syntax/extensions/test_smarty.py
@@ -45,6 +45,18 @@ class TestSmarty(TestCase):
             '<p>&lsquo;Quoted &ldquo;words&rdquo; in a larger quote.&rsquo;</p>'
         )
         self.assertMarkdownRenders(
+            '"Quoted words at the \'end.\'"',
+            '<p>&ldquo;Quoted words at the &lsquo;end.&rsquo;&rdquo;</p>'
+        )
+        self.assertMarkdownRenders(
+            '\'Quoted words at the "end."\'',
+            '<p>&lsquo;Quoted words at the &ldquo;end.&rdquo;&rsquo;</p>'
+        )
+        self.assertMarkdownRenders(
+            '(He replied, "She said \'Hello.\'")',
+            '<p>(He replied, &ldquo;She said &lsquo;Hello.&rsquo;&rdquo;)</p>'
+        )
+        self.assertMarkdownRenders(
             '"quoted" text and **bold "quoted" text**',
             '<p>&ldquo;quoted&rdquo; text and <strong>bold &ldquo;quoted&rdquo; text</strong></p>'
         )

--- a/tests/test_syntax/extensions/test_smarty.py
+++ b/tests/test_syntax/extensions/test_smarty.py
@@ -57,6 +57,10 @@ class TestSmarty(TestCase):
             '<p>(He replied, &ldquo;She said &lsquo;Hello.&rsquo;&rdquo;)</p>'
         )
         self.assertMarkdownRenders(
+            '<span>He replied, "She said \'Hello.\'"</span>',
+            '<p><span>He replied, &ldquo;She said &lsquo;Hello.&rsquo;&rdquo;</span></p>'
+        )
+        self.assertMarkdownRenders(
             '"quoted" text and **bold "quoted" text**',
             '<p>&ldquo;quoted&rdquo; text and <strong>bold &ldquo;quoted&rdquo; text</strong></p>'
         )

